mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala

```diff
@@ -173,6 +173,33 @@ private[spark] class NodeIdCache(
   }
 }
 
+/**
+ * Incrementally assigns new node indices per tree. 
+ * The node index assigner tracks the maximum node index that
+ * has been assigned to a given tree.
+ * 
+ * @param rootNodeIndex Index for the root node
+ */
+private[spark] case class NodeIndexAssigner(rootNodeIndex: Int) {
+  private var maxNodeIndex : Int = rootNodeIndex
+
+  /**
+   * Check if new node indices can be assigned
+   * @return True if new node indices can be assigned
+   */
+  def hasNext : Boolean = maxNodeIndex < Int.MaxValue
+
+  /**
+   * Assign next node index by incrementing the maximum node index for tree
+   * @return Next node index
+   */
+  def nextIndex() : Int = {
```

Comment:
The update is being made by a single thread at the driver and I didn't want to add the overhead of locking. Should I add a comment to add synchronization if used in a multi-threaded program?